ledger/src/blockstore_processor.rs

```diff
@@ -165,44 +182,54 @@ pub fn execute_batch(
         vec![]
     };
 
-    let (commit_results, balances) = batch.bank().load_execute_and_commit_transactions(
-        batch,
-        MAX_PROCESSING_AGE,
-        transaction_status_sender.is_some(),
-        ExecutionRecordingConfig::new_single_setting(transaction_status_sender.is_some()),
-        timings,
-        log_messages_bytes_limit,
-    );
+    let pre_commit_callback = |timings: &mut _, processing_results: &_| {
+        match extra_pre_commit_callback {
+            None => {
+                get_first_error(batch, processing_results)?;
+                check_block_cost_limits_if_enabled(batch, bank, timings, processing_results)?;
+            }
+            Some(extra_pre_commit_callback) => {
+                // We're entering into the block-producing unified scheduler special case...
+                // `processing_results` should always contain exactly only 1 result in that case.
+                assert_eq!(processing_results.len(), 1);
+                assert!(transaction_indexes.is_empty());
+
+                if let Some(index) = extra_pre_commit_callback(&processing_results[0])? {
+                    let transaction_indexes = transaction_indexes.to_mut();
+                    // Adjust the empty new vec with the exact needed capacity. Otherwise, excess
+                    // cap would be reserved on `.push()` in it.
+                    transaction_indexes.reserve_exact(1);
+                    transaction_indexes.push(index);
+                }
+            }
+        }
+        Ok(())
+    };
+
+    let (commit_results, balances) = batch
+        .bank()
+        .load_execute_and_commit_transactions_with_pre_commit_callback(
+            batch,
+            MAX_PROCESSING_AGE,
+            transaction_status_sender.is_some(),
+            ExecutionRecordingConfig::new_single_setting(transaction_status_sender.is_some()),
+            timings,
+            log_messages_bytes_limit,
+            pre_commit_callback,
```

Comment:
this is always present now.
Where else do we call `load_execute_and_commit_transactions` without a callback?

I thought both branches of replay go through this `execute_batch` function?

Reply:
AFAICT it done only in tests/benches, we could make the variant without callback dcou, and clean it up eventually so that those calls default to using the same replay variant of the callback.

Reply:
> AFAICT it done only in tests/benches, we could make the variant without callback dcou,

mostly, yes. i tried to dcou it. but turned out it's not that easy.

as one of blockers, it's used by `banks-server` crate:

![image](https://github.com/user-attachments/assets/df9d767d-ba1b-4744-891d-1e59284fdbb6)

Reply:
> clean it up eventually so that those calls default to using the same replay variant of the callback.

:+1: